prow/gerrit/client/client.go

```diff
@@ -671,7 +671,12 @@ func (h *gerritInstanceHandler) queryChangesForProjectWithoutMetrics(log logrus.
 	seenPos := map[int]int{}
 	add := func(ci gerrit.ChangeInfo) {
 		if p, ok := seenPos[ci.Number]; ok {
+			for ; p < len(pending)-1; p++ {
+				pending[p] = pending[p+1]
+				seenPos[pending[p].Number]--
```

Comment:
@airbornepony  I think `pending[p] = pending[p+1]` is shifting all the entries to the left by 1. And line 676 needs to decrement the value in seenPos because `pending[p]` is actually holding `pending[p+1]`'s ChangeInfo. Finally line 679 assigns the largest value of `p` (position index) to `seenPos[ci.Number]` because we want it to be at the very end (because we just saw it).

While I assume this understanding is correct, I'm afraid I can't say it with confidence because of the manual +/- by-1 things happening. Also, shouldn't it be possible to use slice syntax to do the "move remaining elements to the left by 1"? This might be what Cole's version was doing originally IIRC. While I'm not conversant in slice syntax that style is probably more idiomatic Go.

Overall though I think we should move this `add` func out and unit test it to make sure.

Reply:
I'm working on pulling the function out and adding unit tests.

Reply:
Done (see tip commit).